File: web/src/utils/entities.ts
import { Exocore, exocore, fromProtoTimestamp, MutationBuilder } from 'exocore';
import Emojis from '../utils/emojis';
import { exomind } from '../protos';
import _ from 'lodash';

export class EntityTraits {
    public entity: exocore.store.IEntity;

    private typeTraits: { [type: string]: exocore.store.ITrait[] } = {};
    private idTraits: { [id: string]: exocore.store.ITrait } = {};
    private idInstances: { [id: string]: EntityTrait<unknown> } = {};
    private priorityTraitId?: string;

    constructor(entity: exocore.store.IEntity) {
        this.entity = entity;

        // check if it's a special entity (ex: inbox)
        let priorityTrait: [exocore.store.ITrait, ITraitConstants] = null;
        if (this.entity.id in TRAITS_CONSTANTS) {
            const traitsConstants = TRAITS_CONSTANTS[this.entity.id];
            for (const trait of this.entity.traits) {
                if (trait.id == this.entity.id) {
                    priorityTrait = [trait, traitsConstants];
                    break;
                }
            }
        }

        // index traits by ids and types
        for (const trait of this.entity.traits) {
            const msgType = Exocore.registry.canonicalFullName(trait.message.type_url);
            if (!(msgType in this.typeTraits)) {
                this.typeTraits[msgType] = [trait];
            } else {
                this.typeTraits[msgType].push(trait);
            }

            this.idTraits[trait.id] = trait;

            let traitConstants;
            if (this.entity.id == trait.id && this.entity.id in TRAITS_CONSTANTS) {
                // special entity
                traitConstants = TRAITS_CONSTANTS[this.entity.id];
            } else if (msgType in TRAITS_CONSTANTS) {
                traitConstants = TRAITS_CONSTANTS[msgType];
            }

            if (traitConstants && ((priorityTrait == null || traitConstants.order < priorityTrait[1].order))) {
                priorityTrait = [trait, traitConstants];
            }
        }

        if (priorityTrait != null) {
            this.priorityTraitId = priorityTrait[0].id;
        } else if (this.entity.traits.length > 0) {
            this.priorityTraitId = this.entity.traits[0].id;
        }
    }

    get id(): string {
        return this.entity.id;
    }

    traitOfType<T>(msg: unknown): EntityTrait<T> | null {
        const traits = this.traitsOfType<T>(msg);
        if (traits.length > 0) {
            return traits[0];
        } else {
            return null;
        }
    }

    traitsOfType<T>(msg: unknown): EntityTrait<T>[] {
        const msgType = Exocore.registry.messageFullName(msg);
        return this.entity.traits.flatMap((trait: exocore.store.ITrait) => {
            if (trait.message.type_url.endsWith(msgType)) {
                return [this.trait(trait.id)];
            } else {
                return [];
            }
        });
    }

    trait<T>(id: string): EntityTrait<T> | null {
        const trait = this.idTraits[id];
        if (!trait) {
            return null;
        }

        if (!(id in this.idInstances)) {
            const et = new EntityTrait<unknown>(
                this,
                trait,
                Exocore.registry.unpackAny(trait.message),
            );
            this.idInstances[id] = et;

            return et as EntityTrait<T>;
        }

        return this.idInstances[id] as EntityTrait<T>;
    }

    get priorityTrait(): EntityTrait<unknown> | null {
        if (this.priorityTraitId != null) {
            return this.trait(this.priorityTraitId);
        } else {
            return null;
        }
    }

    priorityMatch(matcher: ITraitMatcher): unknown {
        const priorityTrait = this.priorityTrait;
        if (priorityTrait == null) {
            return null;
        }

        return priorityTrait.match(matcher);
    }
}

export class EntityTrait<T> {
    trait: exocore.store.ITrait;
    message: T;
    et: EntityTraits;

    constructor(et: EntityTraits, trait: exocore.store.ITrait, message: T) {
        this.et = et;
        this.trait = trait;
        this.message = message;
    }

    get id(): string {
        return this.trait.id;
    }

    get constants(): ITraitConstants {
        const msgType = Exocore.registry.messageFullName(this.message);
        if (this.trait.id == this.et.id && this.et.id in TRAITS_CONSTANTS) {
            // special entity
            return TRAITS_CONSTANTS[this.et.id];
        } else if (msgType in TRAITS_CONSTANTS) {
            return TRAITS_CONSTANTS[msgType];
        } else {
            return TRAITS_CONSTANTS['unknown'];
        }
    }

    get creationDate(): Date | null {
        if (this.trait.creationDate != null) {
            return fromProtoTimestamp(this.trait.creationDate);
        } else {
            return null;
        }
    }

    get modificationDate(): Date | null {
        if (this.trait.modificationDate != null) {
            return fromProtoTimestamp(this.trait.modificationDate);
        } else {
            return null;
        }
    }

    get displayName(): string {
        if (this.constants.nameFunction) {
            return this.constants.nameFunction(this.message);
        }

        if (this.constants.name) {
            return this.constants.name;
        }

        if (this.constants.nameField) {
            const dict = this.message as unknown as { [p: string]: string; };
            const name = dict[this.constants.nameField];
            return name ?? this.constants.nameDefault ?? '*UNTITLED*';
        }

        return '*UNTITLED*';
    }

    get icon(): TraitIcon {
        return this.constants.icon(this.message);
    }

    match(matcher: ITraitMatcher): unknown {
        if (this.constants.key == 'exomind.base.v1.EmailThread' && matcher.emailThread) {
            return matcher.emailThread(this);
        } else if (this.constants.key == 'exomind.base.v1.Email' && matcher.email) {
            return matcher.email(this);
        } else if (this.constants.key == 'exomind.base.v1.DraftEmail' && matcher.draftEmail) {
            return matcher.draftEmail(this);
        } else if (this.constants.key == 'exomind.base.v1.Collection' || (this.constants?.collectionLike ?? false) && matcher.collection) {
            return matcher.collection(this);
        } else if (this.constants.key == 'exomind.base.v1.Task' && matcher.task) {
            return matcher.task(this);
        } else if (this.constants.key == 'exomind.base.v1.Note' && matcher.note) {
            return matcher.note(this);
        } else if (this.constants.key == 'exomind.base.v1.Link' && matcher.link) {
            return matcher.link(this);
        } else if (matcher.default) {
            return matcher.default();
        }
    }

    get canEditName(): boolean {
        return !!this.constants.rename;
    }

    get editableName(): string {
        if (this.constants.renameValue) {
            return this.constants.renameValue(this.message);
        } else {
            return this.displayName;
        }
    }

    async rename(newName: string): Promise<exocore.store.IMutationResult> {
        if (!this.constants.rename) {
            return;
        }


        const newMsg = _.clone(this.message);
        this.constants.rename(newMsg, newName);

        await Exocore.store.mutate(
            MutationBuilder
                .updateEntity(this.et.id)
                .putTrait(newMsg, this.id)
                .build()
        );
    }
}

export function isEntityTraits(object: unknown): object is EntityTraits {
    return object instanceof EntityTraits;
}

export interface ITraitMatcher {
    emailThread?: (trait: EntityTrait<exomind.base.v1.IEmailThread>) => unknown;
    email?: (trait: EntityTrait<exomind.base.v1.IEmail>) => unknown;
    draftEmail?: (trait: EntityTrait<exomind.base.v1.IDraftEmail>) => unknown;
    collection?: (trait: EntityTrait<exomind.base.v1.ICollection>) => unknown;
    task?: (trait: EntityTrait<exomind.base.v1.ITask>) => unknown;
    note?: (trait: EntityTrait<exomind.base.v1.INote>) => unknown;
    link?: (trait: EntityTrait<exomind.base.v1.ILink>) => unknown;
    default?: () => unknown;
}

export interface ITraitConstants {
    key: string;
    name?: string;
    nameField?: string;
    nameDefault?: string;
    nameFunction?: (trait: unknown) => string;
    icon: (trait: unknown) => TraitIcon;
    color: number;
    order: number;
    collectionLike?: boolean;
    renameValue?: (trait: unknown) => string;
    rename?: (trait: unknown, newName: string) => void;
}

export type TraitIcon = { fa: string } | { emoji: string };

export const TRAITS_CONSTANTS: { [type: string]: ITraitConstants } = {
    'inbox': {
        key: 'inbox',
        name: 'Inbox',
        icon: () => { return { fa: 'inbox' }; },
        collectionLike: true,
        color: 4,
        order: 0
    },
    'favorites': {
        key: 'favorites',
        name: 'Favorites',
        icon: () => { return { fa: 'star' }; },
        collectionLike: true,
        color: 4,
        order: 1
    },
    'exomind.base.v1.EmailThread': {
        key: 'exomind.base.v1.EmailThread',
        nameField: 'subject',
        nameDefault: 'Untitled email',
        icon: () => { return { fa: 'envelope-o' }; },
        color: 1,
        order: 2
    },
    'exomind.base.v1.DraftEmail': {
        key: 'exomind.base.v1.DraftEmail',
        nameField: 'subject',
        nameDefault: 'Untitled email',
        icon: () => { return { fa: 'envelope-o' }; },
        color: 6,
        order: 3
    },
    'exomind.base.v1.Email': {
        key: 'exomind.base.v1.Email',
        nameField: 'subject',
        nameDefault: 'Untitled email',
        icon: () => { return { fa: 'envelope-o' }; },
        color: 6,
        order: 4
    },
    'exomind.base.v1.Collection': {
        key: 'exomind.base.v1.Collection',
        nameFunction: (trait) => {
            const col = trait as exomind.base.v1.ICollection;

            if (col.name) {
                if (Emojis.hasEmojiPrefix(col.name)) {
                    const [, title] = Emojis.extractEmojiPrefix(col.name);
                    return title;
                } else {
                    return col.name;
                }
            } else {
                return 'Untitled collection';
            }
        },
        icon: (trait) => {
            const col = trait as exomind.base.v1.ICollection;
            if (Emojis.hasEmojiPrefix(col.name)) {
                const [emoji] = Emojis.extractEmojiPrefix(col.name);
                return { emoji };
            } else {
                return { fa: 'folder-o' };
            }
        },
        color: 2,
        order: 5,
        collectionLike: true,
        renameValue: (trait: unknown) => {
            const col = trait as exomind.base.v1.ICollection;
            return col.name;
        },
        rename: (trait: unknown, newName: string): void => {
            const collection = trait as exomind.base.v1.ICollection;
            collection.name = newName;
        },
    },
    'exomind.base.v1.Task': {
        key: 'exomind.base.v1.Task',
        nameField: 'title',
        nameDefault: 'Untitled task',
        icon: () => { return { fa: 'check-square-o' }; },
        color: 7,
        order: 6,
        rename: (entity: unknown, newName: string): void => {
            const task = entity as exomind.base.v1.ITask;
            task.title = newName;
        },
    },
    'exomind.base.v1.Note': {
        key: 'exomind.base.v1.Note',
        nameField: 'title',
        nameDefault: 'Untitled note',
        icon: () => { return { fa: 'pencil' }; },
        color: 3,
        order: 7,
        rename: (entity: unknown, newName: string): void => {
            const note = entity as exomind.base.v1.INote;
            note.title = newName;
        },
    },
    'exomind.base.v1.Link': {
        key: 'exomind.base.v1.Link',
        nameField: 'title',
        nameDefault: 'Untitled link',
        icon: () => { return { fa: 'link' }; },
        color: 9,
        order: 8
    },
    'unknown': {
        key: 'unknown',
        nameField: '*UNKNOWN*',
        icon: () => { return { fa: 'question' }; },
        color: 0,
        order: 9
    }
};

File: exomind/ios/Exomind/Entities.swift
import Foundation
import Exocore
import SwiftProtobuf

typealias EntityId = String
typealias TraitId = String

extension Exocore_Store_Entity {
    func toExtension() -> EntityExt {
        EntityExt(entity: self)
    }
}

class EntityExt {
    let inner: Exocore_Store_Entity;

    private let idTraits: [TraitId: Exocore_Store_Trait]
    private let typeTraits: [TraitId: [Exocore_Store_Trait]]
    private var traitInstances: [TraitId: Message] = [:]
    private let _priorityTrait: Exocore_Store_Trait?;

    var creationDate: Date
    var modificationDate: Date?
    var anyDate: Date

    init(entity: Exocore_Store_Entity) {
        self.inner = entity
        self.anyDate = Date()

        var priorityTrait: (Exocore_Store_Trait, TraitConstants)?

        // check if it's a special entity (ex: inbox)
        if let entityConstants = TraitsConstants[entity.id] {
            for trait in entity.traits {
                if trait.id == entity.id {
                    priorityTrait = (trait, entityConstants)
                    break
                }
            }
        }

        // index traits by ids and types
        var idTraits: [String: Exocore_Store_Trait] = [:]
        var typeTraits: [String: [Exocore_Store_Trait]] = [:]
        for trait in entity.traits {
            idTraits[trait.id] = trait

            let traitType = trait.canonicalFullName()
            var traits = typeTraits[traitType] ?? []
            traits.append(trait)
            typeTraits[traitType] = traits

            let traitConstants: TraitConstants?;
            if let entityConsts = TraitsConstants[entity.id], entity.id == trait.id {
                traitConstants = entityConsts
            } else {
                traitConstants = TraitsConstants[traitType]
            }

            if let traitConstants = traitConstants, (priorityTrait == nil || traitConstants.order < priorityTrait!.1.order) {
                priorityTrait = (trait, traitConstants)
            }
        }
        self.idTraits = idTraits
        self.typeTraits = typeTraits
        self._priorityTrait = priorityTrait.map({ $0.0 })
        self.creationDate = entity.hasCreationDate ? entity.creationDate.date : Date()
        self.modificationDate = entity.hasModificationDate ? entity.modificationDate.date : nil
        self.anyDate = self.modificationDate ?? self.creationDate
    }

    var id: EntityId {
        get {
            self.inner.id
        }
    }

    func trait<T: Message>(withId id: TraitId) -> TraitInstance<T>? {
        guard let trait = self.idTraits[id] else {
            return nil
        }

        if let message = self.traitInstances[id] as? T {
            return TraitInstance(entity: self, trait: trait, message: message)
        }

        guard let message = try? T.init(unpackingAny: trait.message) else {
            return nil
        }
        self.traitInstances[id] = message

        return TraitInstance(entity: self, trait: trait, message: message)
    }

    func trait(anyWithId id: TraitId) -> AnyTraitInstance? {
        guard let trait = self.idTraits[id],
              let traitConstants = TraitsConstants[trait.canonicalFullName()] else {
            return nil
        }

        switch traitConstants.traitType {
        case .inbox:
            let trait: TraitInstance<Exomind_Base_V1_Collection>? = self.trait(withId: trait.id)
            return trait
        case .favorites:
            let trait: TraitInstance<Exomind_Base_V1_Collection>? = self.trait(withId: trait.id)
            return trait
        case .emailThread:
            let trait: TraitInstance<Exomind_Base_V1_EmailThread>? = self.trait(withId: trait.id)
            return trait
        case .draftEmail:
            let trait: TraitInstance<Exomind_Base_V1_DraftEmail>? = self.trait(withId: trait.id)
            return trait
        case .email:
            let trait: TraitInstance<Exomind_Base_V1_Email>? = self.trait(withId: trait.id)
            return trait
        case .collection:
            let trait: TraitInstance<Exomind_Base_V1_Collection>? = self.trait(withId: trait.id)
            return trait
        case .task:
            let trait: TraitInstance<Exomind_Base_V1_Task>? = self.trait(withId: trait.id)
            return trait
        case .note:
            let trait: TraitInstance<Exomind_Base_V1_Note>? = self.trait(withId: trait.id)
            return trait
        case .link:
            let trait: TraitInstance<Exomind_Base_V1_Link>? = self.trait(withId: trait.id)
            return trait
        }
    }

    func traitOfType<T: Message>(_ message: T.Type) -> TraitInstance<T>? {
        let traits = self.typeTraits[message.protoMessageName] ?? []
        return traits.compactMap({ (trait) -> TraitInstance<T>? in
                    self.trait(withId: trait.id)
                })
                .first
    }

    func traitsOfType<T: Message>(_ message: T.Type) -> [TraitInstance<T>] {
        let traits = self.typeTraits[message.protoMessageName] ?? []
        return traits.compactMap({ (trait) -> TraitInstance<T>? in
            self.trait(withId: trait.id)
        })
    }

    lazy var priorityTrait: AnyTraitInstance? = {
        guard let trait = self._priorityTrait else {
            return nil
        }

        return self.trait(anyWithId: trait.id)
    }()
}

extension EntityExt: Equatable {
    static func ==(lhs: EntityExt, rhs: EntityExt) -> Bool {
        lhs.id == rhs.id && lhs.anyDate == rhs.anyDate
    }
}

protocol AnyTraitInstance {
    var entity: EntityExt? { get }
    var id: TraitId { get }
    var trait: Exocore_Store_Trait { get }
    var constants: TraitConstants? { get }
    var type: TraitType? { get }
    var displayName: String { get }
    var strippedDisplayName: String { get }
    var creationDate: Date { get }
    var modificationDate: Date? { get }

    func typeInstance() -> TraitTypeInstance?
}

struct TraitInstance<T: Message>: AnyTraitInstance {
    weak var entity: EntityExt?;
    let trait: Exocore_Store_Trait
    let message: T
    let displayName: String
    let constants: TraitConstants?

    init(entity: EntityExt, trait: Exocore_Store_Trait, message: T) {
        self.entity = entity
        self.trait = trait
        self.message = message
        self.constants = getTraitConstants(entity: entity, trait: trait)

        if let constants = self.constants {
            self.displayName = TraitInstance.getDisplayName(constants: constants, message: message)
        } else {
            self.displayName = "*UNKNOWN*"
        }
    }

    var id: TraitId {
        get {
            self.trait.id
        }
    }

    var creationDate: Date {
        get {
            self.trait.creationDate.date
        }
    }

    var modificationDate: Date? {
        get {
            if self.trait.hasModificationDate {
                return self.trait.modificationDate.date
            }

            return nil
        }
    }

    var anyDate: Date {
        get {
            self.modificationDate ?? self.creationDate
        }
    }

    var type: TraitType? {
        get {
            self.constants?.traitType
        }
    }

    func typeInstance() -> TraitTypeInstance? {
        guard let constants = self.constants else {
            return nil
        }

        switch constants.traitType {
        case .inbox:
            return .inbox(trait: self as! TraitInstance<Exomind_Base_V1_Collection>)
        case .favorites:
            return .favorites(trait: self as! TraitInstance<Exomind_Base_V1_Collection>)
        case .emailThread:
            return .emailThread(trait: self as! TraitInstance<Exomind_Base_V1_EmailThread>)
        case .draftEmail:
            return .draftEmail(trait: self as! TraitInstance<Exomind_Base_V1_DraftEmail>)
        case .email:
            return .email(trait: self as! TraitInstance<Exomind_Base_V1_Email>)
        case .collection:
            return .collection(trait: self as! TraitInstance<Exomind_Base_V1_Collection>)
        case .task:
            return .task(trait: self as! TraitInstance<Exomind_Base_V1_Task>)
        case .note:
            return .note(trait: self as! TraitInstance<Exomind_Base_V1_Note>)
        case .link:
            return .link(trait: self as! TraitInstance<Exomind_Base_V1_Link>)
        }
    }

    func toAny() -> AnyTraitInstance? {
        self.entity?.trait(anyWithId: self.id)
    }

    var strippedDisplayName: String {
        get {
            if let constants = self.constants {
                return TraitInstance.getDisplayName(constants: constants, message: message, strip: true)
            } else {
                return "*UNKNOWN*"
            }
        }
    }

    private static func getDisplayName<M: Message>(constants: TraitConstants, message: M, strip: Bool = false) -> String {
        if let name = constants.name {
            return name
        }

        var name: String?;
        switch message {
        case let emailThread as Exomind_Base_V1_EmailThread:
            name = emailThread.subject.nonEmpty() ?? "Untitled email"
        case let draftEmail as Exomind_Base_V1_DraftEmail:
            name = draftEmail.subject.nonEmpty() ?? "Untitled email"
        case let email as Exomind_Base_V1_Email:
            name = email.subject.nonEmpty() ?? "Untitled email"
        case let collection as Exomind_Base_V1_Collection:
            if strip && collection.name.startsWithEmoji() {
                let (_, rest) = collection.name.splitFirstEmoji()
                name = rest.nonEmpty() ?? "Untitled collection"
            } else {
                name = collection.name.nonEmpty() ?? "Untitled collection"
            }
        case let task as Exomind_Base_V1_Task:
            name = task.title.nonEmpty() ?? "Untitled task"
        case let note as Exomind_Base_V1_Note:
            name = note.title.nonEmpty() ?? "Untitled note"
        case let link as Exomind_Base_V1_Link:
            name = link.title.nonEmpty() ?? "Untitled link"
        default:
            name = nil
        }

        return name ?? constants.nameDefault ?? "*UNKNOWN*"
    }
}

enum TraitType {
    case inbox
    case favorites
    case emailThread
    case draftEmail
    case email
    case collection
    case task
    case note
    case link
}

enum TraitTypeInstance {
    case inbox(trait: TraitInstance<Exomind_Base_V1_Collection>)
    case favorites(trait: TraitInstance<Exomind_Base_V1_Collection>)
    case emailThread(trait: TraitInstance<Exomind_Base_V1_EmailThread>)
    case draftEmail(trait: TraitInstance<Exomind_Base_V1_DraftEmail>)
    case email(trait: TraitInstance<Exomind_Base_V1_Email>)
    case collection(trait: TraitInstance<Exomind_Base_V1_Collection>)
    case task(trait: TraitInstance<Exomind_Base_V1_Task>)
    case note(trait: TraitInstance<Exomind_Base_V1_Note>)
    case link(trait: TraitInstance<Exomind_Base_V1_Link>)
}

struct TraitConstants {
    let key: String
    let traitType: TraitType
    let name: String?
    let nameDefault: String?
    let icon: String
    let color: Int
    let order: Int
    let collectionLike: Bool
    let canPreview: Bool
}

func getTraitConstants(entity: EntityExt, trait: Exocore_Store_Trait) -> TraitConstants? {
    if let entityConsts = TraitsConstants[entity.id], entity.id == trait.id {
        return entityConsts
    } else {
        return TraitsConstants[trait.canonicalFullName()]
    }
}

let TraitsConstants: [String: TraitConstants] = [
    "inbox": TraitConstants(
            key: "inbox",
            traitType: .inbox,
            name: "Inbox",
            nameDefault: nil,
            icon: "inbox",
            color: 4,
            order: 0,
            collectionLike: true,
            canPreview: false
    ),
    "favorites": TraitConstants(
            key: "favorites",
            traitType: .favorites,
            name: "Favorites",
            nameDefault: nil,
            icon: "star",
            color: 4,
            order: 1,
            collectionLike: true,
            canPreview: false
    ),
    "exomind.base.v1.EmailThread": TraitConstants(
            key: "exomind.base.v1.EmailThread",
            traitType: .emailThread,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 1,
            order: 2,
            collectionLike: false,
            canPreview: true
    ),
    "exomind.base.v1.DraftEmail": TraitConstants(
            key: "exomind.base.v1.DraftEmail",
            traitType: .draftEmail,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 6,
            order: 3,
            collectionLike: false,
            canPreview: true
    ),
    "exomind.base.v1.Email": TraitConstants(
            key: "exomind.base.v1.Email",
            traitType: .email,
            name: nil,
            nameDefault: "Untitled email",
            icon: "envelope-o",
            color: 6,
            order: 4,
            collectionLike: false,
            canPreview: true
    ),
    "exomind.base.v1.Collection": TraitConstants(
            key: "exomind.base.v1.Collection",
            traitType: .collection,
            name: nil,
            nameDefault: nil,
            icon: "folder-o",
            color: 2,
            order: 5,
            collectionLike: true,
            canPreview: true
    ),
    "exomind.base.v1.Task": TraitConstants(
            key: "exomind.base.v1.Task",
            traitType: .task,
            name: nil,
            nameDefault: nil,
            icon: "check-square-o",
            color: 7,
            order: 6,
            collectionLike: false,
            canPreview: false
    ),
    "exomind.base.v1.Note": TraitConstants(
            key: "exomind.base.v1.Note",
            traitType: .note,
            name: nil,
            nameDefault: nil,
            icon: "pencil",
            color: 3,
            order: 7,
            collectionLike: false,
            canPreview: true
    ),
    "exomind.base.v1.Link": TraitConstants(
            key: "exomind.base.v1.Link",
            traitType: .link,
            name: nil,
            nameDefault: "Untitled link",
            icon: "link",
            color: 9,
            order: 8,
            collectionLike: false,
            canPreview: true
    )
]
